server/s3_motr_reader.cc

```diff
@@ -344,6 +344,39 @@ size_t S3MotrReader::get_next_block(char **data) {
   *data = (char *)motr_rw_op_context->data->ov_buf[iteration_index];
   data_read = motr_rw_op_context->data->ov_vec.v_count[iteration_index];
   iteration_index++;
+
+  size_t length = data_read;
+  if (total_size_to_read < total_size_read + length)
+    length = total_size_to_read - total_size_read;
+  if (multipart_part_size == 0) {
+    // non-multipart-upload case. Just calculate md5 of the entire object.
+    md5crypt.Update(*data, length);
+  } else {
+    // the object was created with multipart upload.
+    // Calculate md5 checksums for each part independently.
+    size_t remaining = length;
+    // Each iteration jumps to the next part or the end of the buffer,
+    // whichever comes first.
+    while (remaining > 0) {
+      // current absolute position in object for the purpose of checksumming
+      size_t pos = total_size_read + length - remaining;
+      size_t next_part_pos;
+      if (pos / multipart_part_size >= multipart_num_of_parts - 1)
+        next_part_pos = total_size_to_read;
+      else
+        next_part_pos = (pos / multipart_part_size + 1) * multipart_part_size;
```

Comment:
This logic also challenging.

Reply:
Added a few comments. Please let me know if the logic is still not clear.

Reply:
I'm still not following.  Maybe if you created some temporary variables.  Like

foo = pos / multipart\_part\_size; // where foo is some intuitive variable name

bar = multipart\_num\_of\_parts - 1; // where bar is some intuitive variable name

then

if ( foo >= bar) {

.....

Reply:
Added more comments and changed `if` condition to make it even more easier to read. Please re-check.

Reply:
Much better.  Thank you.  @knrajnambiar76 , does this work for you?